File: w2015/hw1/q1_people_you_may_know_spark.py
# execfile('friend_you_may_know.py')

#fileName = 'q1testdata.txt'
#from pyspark import SparkContext, SparkConf
# conf = SparkConf()
# conf.setMaster("local")
# conf.setAppName("Friends you may know")
# conf.set("spark.executor.memory", "16g")
#sc = SparkContext(conf=conf)
from collections import Counter
import itertools

fileName = 'soc-LiveJournal1Adj.txt'
N = 10  #only ouput 10 most possible friends

def connecteds_and_commons(line):
  minimum = -9999999999
  user, friends = line.split('\t')
  friends = friends.split(',')
  connecteds = [((user, friend), minimum) for friend in friends]
  commons = [(pair, 1) for pair in itertools.permutations(friends, 2)]
  return connecteds + commons

friendsListRDD = (sc
                  .textFile( fileName, 16 )
                  .flatMap( connecteds_and_commons )
                  .reduceByKey( lambda total, current: total + current )
                  .filter(lambda (pair, counts): counts > 0)
                  .map(lambda ((user, friend), counts): (user, (counts, friend)))
                  .groupByKey()
                  .map(lambda (user, suggestions):(user, Counter( dict( (friend, count) for count, friend in suggestions ) ).most_common( N ) ) )
                  #.cache()
                   )


print "924"
print friendsListRDD.lookup('924')
#print friendsListRDD.collect()

File: w2015/hw1/q1_people_you_may_know_spark.rb
# in jenv exec ruby-spark shell
# load('q1_people_you_may_know_spark.rb')

require 'ruby-spark'

parse_friends_line = lambda do |line|
  minimum, friends = -9999999999, nil
  user, friends = line.split("\t")
  if line.nil? || friends.nil?  #somebody got no friends...
    [ [user, nil] ]
  else
    friends = friends.split(",")
    connected = friends.map{ |friend| [[user, friend], minimum ] }
    commons = friends.permutation(2).map{ |pair| [pair, 1] }
    connected + commons
  end
end

file_name = 'soc-LiveJournal1Adj.txt'
#file_name = 'q1testdata.txt'

friends_list_RDD = $sc
  .textFile( file_name )
  .flatMap( parse_friends_line )
  .reduceByKey( lambda{ |total, current| total + current } )
  .filter( lambda{ |pair, counts| !counts.nil? && counts > 0 } )
  .map( lambda{|pair, counts| [pair.first, [counts, pair.last]] } )
  .groupByKey
  .map(lambda{|user, suggestions| [user, suggestions.sort.reverse ] }  )
# .cache

#print friends_list_RDD.lookup('924')
print friends_list_RDD.take(2)
#complain method missing...

#puts "924"
#print friendsListRDD.lookup('924')  #Need Ruby-Spark Aug 16 2015 version
# should be 439,2409,6995,11860,15416,43748,45881
